### PR TITLE
Fix #1247: Hints & Solutions support in Question Player

### LINKED ISSUE #1247: Hints & Solutions support in Question Player [BLOCKED on #500]
Question Player needs support for Hints & Solutions.

It will done in the same way as ExplorationPlayer (State Fragment). 

Mock Link: https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/12c7159f-4c16-4c89-8311-59963c28c1e4/LM-Q5-HS-Hint-Dialog-Close-Hint-1-

### PR DESCRIPTION
**Pull request by veena14cs**
## Explanation
This PR tracks the issue #1247. It implements hints and solution support for question player.

## Design Doc
https://docs.google.com/document/d/1NrZOTcVXhJxwCG_3nGcp47jBHLufLRrU5Kcv9aVjj0M/edit?usp=sharing

## Fuctionality
The first hint is unlocked after 60s and subsequent hints are unlocked at 30s intervals. The time interval for hints to be unlocked are accelerated if the student submits wrong answers (ie. 2 wrong answers releases the next hint in 10s).

**Mock Link**: https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/12c7159f-4c16-4c89-8311-59963c28c1e4/LM-Q5-HS-Hint-Dialog-Close-Hint-1-

## Screenshot
Screenshot_20200626-152457 (2)

Screenshot_20200626-152529
Screenshot_20200626-152537

### CODE REVIEW

File: app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt]: Please remove this logger if not needed.

File: app/src/main/res/layout/question_player_fragment.xml
Comment (rt4914) [inline app/src/main/res/layout/question_player_fragment.xml]: This can be shifted to above line. Ditto elsewhere in this file.

File: app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt]: Not sure, if this is correct or not, meaning if we should depend on its parent or not. While having conversation with @BenHenning he mentioned that the `StatePlayerRecyclerViewAssembler` should be improvised in such a way that it does not depend on question player or exploration player and I was thinking if that applies here also. 
Will assign this to @BenHenning for better review on approach.

**Review by rt4914 — CHANGES_REQUESTED** [review]
LGTM, but I have following concerns:
1. Domain layer tests are missing
2. App layer tests are missing, if this is blocked on something than file an issue and mention it in this PR accordingly.
3. Will defer to @BenHenning for approach confirmation.
4. Also, please add steps/screenshots to make it easier for other reviewers.

File: app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt:26]: Will this work with questions?

File: app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt
Comment (BenHenning) [reply app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt]: Thanks Rajat. I think we want to avoid specializing this fragment to each place it can be located, but I can see why it's challenging in this case.

One alternative might be to move the common hints & solutions logic to a common fragment, and have wrapper fragments for explorations & questions that keep things specialized. Consider this fragment hierarchy:

- ExplorationFragment
  - HintsAndSolutionExplorationManagerFragment
    - HintsAndSolutionDialogFragment

and


- QuestionFragment
  - HintsAndSolutionQuestionManagerFragment
    - HintsAndSolutionDialogFragment

Thoughts?

File: app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt:26]: @BenHenning Yes its is working.

File: app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt]: Yeah this can be done, if we want to have specific files for each feature.

File: app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt]: Removed.

File: app/src/main/res/layout/question_player_fragment.xml
Comment (veena14cs) [reply app/src/main/res/layout/question_player_fragment.xml]: Done.

File: app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragmentPresenter.kt]: > Thanks Rajat. I think we want to avoid specializing this fragment to each place it can be located, but I can see why it's challenging in this case.
> 
> One alternative might be to move the common hints & solutions logic to a common fragment, and have wrapper fragments for explorations & questions that keep things specialized. Consider this fragment hierarchy:
> 
> * ExplorationFragment
>   
>   * HintsAndSolutionExplorationManagerFragment
>     
>     * HintsAndSolutionDialogFragment
> 
> and
> 
> * QuestionFragment
>   
>   * HintsAndSolutionQuestionManagerFragment
>     
>     * HintsAndSolutionDialogFragment
> 
> Thoughts?

@BenHenning @veena14cs This approach sounds correct to me, as it will still keep the major logic portion independent of its parent. Thanks.

File: app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragment.kt
Comment (nikitamarysolomanpvt) [inline app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragment.kt]: Please update these as per meeting suggestion...

File: app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionExplorationManagerFragment.kt
Comment (nikitamarysolomanpvt) [inline app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionExplorationManagerFragment.kt]: Please update these as per meeting suggestions...(here  and elsewhere) PTAL on https://docs.google.com/document/d/17mjInnMroI17jlSkn0GhDo-iVw9heF1z8FtjY_saUl0/edit

**Review by nikitamarysolomanpvt — CHANGES_REQUESTED** [review]
Please refer https://docs.google.com/document/d/17mjInnMroI17jlSkn0GhDo-iVw9heF1z8FtjY_saUl0/edit and update KEYs and its implementations.

File: app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragment.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionFragment.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionExplorationManagerFragment.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionExplorationManagerFragment.kt]: Done.

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:165]: Nit:
```
  override fun onExplorationStateLoaded(state: State) {
    this.state = state
  }
```

File: app/src/main/java/org/oppia/app/player/exploration/HintsAndSolutionExplorationManagerFragment.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/exploration/HintsAndSolutionExplorationManagerFragment.kt]: Remove extra blank lines from this file here and above

File: app/src/main/java/org/oppia/app/topic/questionplayer/HintsAndSolutionQuestionManagerFragment.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/topic/questionplayer/HintsAndSolutionQuestionManagerFragment.kt:28]: Remove this extra blank line

File: app/src/main/java/org/oppia/app/topic/questionplayer/HintsAndSolutionQuestionManagerListener.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/topic/questionplayer/HintsAndSolutionQuestionManagerListener.kt:7]: Remove this extra blank line

File: app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivity.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivity.kt]: Remove this extra blank line

File: app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt]: Change this logger tag

File: domain/src/main/assets/sample_questions.json
Comment (rt4914) [inline domain/src/main/assets/sample_questions.json:159]: Do you still need this?

File: domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt
Comment (rt4914) [inline domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt]: Mention what is this `true` and `0` index is for, here and everywhere else in this file.

**Review by rt4914 — CHANGES_REQUESTED** [review]
LGTM. Suggested some nit changes.

Also, there should be some app layer test cases for this PR. If they are blocked on some PR/issue, mention that here.

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:165]: Done.

File: app/src/main/java/org/oppia/app/player/exploration/HintsAndSolutionExplorationManagerFragment.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/exploration/HintsAndSolutionExplorationManagerFragment.kt]: Done.

File: app/src/main/java/org/oppia/app/topic/questionplayer/HintsAndSolutionQuestionManagerFragment.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/topic/questionplayer/HintsAndSolutionQuestionManagerFragment.kt:28]: Done.

File: app/src/main/java/org/oppia/app/topic/questionplayer/HintsAndSolutionQuestionManagerListener.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/topic/questionplayer/HintsAndSolutionQuestionManagerListener.kt:7]: Done.

File: app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivity.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivity.kt]: Done.

File: app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt]: Done.

File: domain/src/main/assets/sample_questions.json
Comment (veena14cs) [reply domain/src/main/assets/sample_questions.json:159]: Yes. There is no solution in json file. So it was added.

File: domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt
Comment (veena14cs) [reply domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt]: Done.

**Comment by veena14cs** [comment]
> LGTM. Suggested some nit changes.
> 
> Also, there should be some app layer test cases for this PR. If they are blocked on some PR/issue, mention that here.

I have mentioned issue number in Test class that tracks testcases for hints and solution.

File: app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt]: @veena14cs PTAL at this Logger TAG too.

File: domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt
Comment (rt4914) [inline domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt]: Nit: `/* hintIndex= */ 0`

**Review by rt4914 — APPROVED** [review]
Minor nit suggested.

File: app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt]: Done.

File: domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt
Comment (veena14cs) [reply domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt]: Done.

**Review by nikitamarysolomanpvt — APPROVED** [review]
LGTM

**Pull request merged**
